### PR TITLE
Challenge 4

### PR DESCRIPTION
**Pull request by nobeeakon**
# Mini-Challenge 4: State Management

## The Challenge
1. Create the Provider for the Global Context.
2. **Consume the Global Context inside the components that require it:** Remember the hook `useContext`? It's time to make use of it!
3. Create test cases for your Global Context and for the components that consume it.

> **Note:* Remember to apply TDD as much as you can.*

## Acceptance Criteria
- [X] The search term is stored and retrieved from the Global Context correctly.
- [X] The appearance theme is stored on the Global Context and applied correctly to the App UI:  _(Didn't store the themes directly in the GlobalContext, since I'm using styledComponents and it is neater to have the themes in a ThemeProvider. However, I stored in the GlobalContext a boolean to toogle themes)_.
- [X] `useReducer` hook is implemented correctly to manage the Global State.

## Bonus Points
- [X] Testing coverage is above 70%. (Please include a screenshot of the code coverage report in your PR description).

imagen

### CODE REVIEW

File: src/components/NavBar/NavBar.spec.jsx [hunk 1]
```diff
@@ -1,52 +1,105 @@
...
+import GlobalContextProvider from '../../providers/Global';
 import AuthProvider from '../../providers/Auth';
 
-import { storage } from '../../utils/storage';
+import { ROUTES } from '../../utils/functions/routes';
 
-import { DARK_MODE_STORAGE_KEY } from '../../utils/constants';
+// TODO: check if mocking history is a good aproach.
+//       It seems to me (Daniel) as an implementation detail...
+const mockHistoryPush = jest.fn();
```
Comment (vmcruz) [inline src/components/NavBar/NavBar.spec.jsx:13]: I think it is ok to test if the `history.push` was called or not. During refactors, the route where it redirects may change, or even break. Testing where it goes helps you to detect these type of issues. You don't want to test if it actually goes to the given route, just if it was called with the expected path.

File: src/components/NavBar/NavBar.spec.jsx [hunk 2]
```diff
@@ -1,52 +1,105 @@
...
-        <DarThemeProvider>
-          <NavBar />
-        </DarThemeProvider>
-      </AuthProvider>
-    );
-
-    const darkModeToggle = getByTestId('toggleTheme-button');
-    act(() => {
-      fireEvent.click(darkModeToggle);
+  // TODO: how to  test toggle dark mode, without testing the actual implementation?
```
Comment (vmcruz) [inline src/components/NavBar/NavBar.spec.jsx]: Ping me next week if you need help with this. But basically, you'll need to mock the `dispatch` and assert that it was called with the `Toggle dark` action. Then in your `reducer` test, you need to assert that the action actually mutates the state to the expected result.

File: src/components/VideoList/VideoListCard/VideoListCard.spec.jsx [hunk 3]
```diff
@@ -128,7 +128,7 @@ describe('Testing VideoCard.component', () => {
                 description={codedString}
               />
             </ThemeProvider>
-          </DarkModeProvider>
+          </GlobalContextProvider>
```
Comment (vmcruz) [inline src/components/VideoList/VideoListCard/VideoListCard.spec.jsx]: Most of your tests were refactored to include the `GlobalContextProvider`. There's an option with `react-testing-library` to create a `custom render` that will wrap your component with all the providers needed. This way, if you need to add a new provider in the future, you won't need to go test by test to add it. Check the following documentation and let me know if you need any help setting it up, we can do a pair programming if needed:

https://testing-library.com/docs/react-testing-library/setup/#custom-render

File: src/components/NavBar/NavBar.component.jsx [hunk 4]
```diff
@@ -26,12 +30,13 @@ import {
 import AvatarImg from '../Avatar';
 
 const NavBar = () => {
+  const history = useHistory();
   const [searchString, setSearchString] = useState('');
 
-  const { toggleDarkMode, isDarkMode } = useDark();
+  const { globalState, dispatchGlobal } = useGlobalContext();
```
Comment (vmcruz) [inline src/components/NavBar/NavBar.component.jsx]: Here you could use your `globalState` to set the initial value for your `useState` hook.

File: src/pages/Results/Results.page.jsx [hunk 5]
```diff
@@ -1,24 +1,26 @@
...
-  const parsedQuery = queryString.parse(location.search);
-  const searchQuery = parsedQuery.search_query;
+const Results = () => {
+  const { globalState } = useGlobalContext();
+  const { searchTerm } = globalState;
 
-  if (!searchQuery) {
+  if (!searchTerm) {
     return <Redirect to="/" />;
   }
```
Comment (vmcruz) [inline src/pages/Results/Results.page.jsx:17]: I feel like this logic could live inside `VideoList`, so you don't need to pass the `searchTerm` as a prop to it. I might be wrong, but consider the possibility.

File: src/providers/Global/Global.provider.jsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+      };
+    case ACTIONS.UPDATE_SEARCH:
+      return {
+        ...state,
+        searchTerm: action.payload,
+      };
+    default:
+      throw new Error(`Global reducer action.type ${action.type} not allowed`);
+  }
+};
```
Comment (vmcruz) [inline src/providers/Global/Global.provider.jsx]: I suggest you to extract the reducer logic to its own file, so you can test it isolated.

**Review by vmcruz — COMMENTED** [review]
Awesome work. Left a couple of comments but overall good work. Keep it up 🚀 .

Edit:
Forgot to mention something. Consider adding the `cursor: pointer` style to the elements that are clickable, like the theme toggle, the home icon, login button, etc.

**Comment by nobeeakon** [comment]
updated branch, taking into account Victor's feedback

**Pull request merged**